Bjorncs/tls certificate validation

bjorncs opened the pull request.

bjorncs committed:
Add data model for AuthorizedPeers

bjorncs committed:
Rewrite JSON serialization of TransportSecurityOptions

 - Use Jackson data bindings on TransportSecurityOptionsEntity
 - Add serialization to JSON
 - Add AuthorizedPeers to TransportSecurityOptions

bjorncs committed:
Use TransportSecurityOptions.Builder

vekterli reviewed (approved):
:+1: just a few nitpicks/questions

bjorncs committed:
Misc minor code improvements

- Improve naming of some methods and parameters
- Add missing trailing newline
- Use missingFieldException(String)
- Remove dead code

bjorncs commented on jrt/src/com/yahoo/jrt/TlsCryptoEngine.java:
```diff
@@ -40,9 +34,10 @@ public TlsCryptoSocket createCryptoSocket(SocketChannel channel, boolean isServe
+        SslContextBuilder builder = new SslContextBuilder();
+        options.getCertificatesFile()
+                .ifPresent(certificates -> builder.withKeyStore(options.getPrivateKeyFile().get(), certificates));
+        options.getCaCertificatesFile().ifPresent(builder::withTrustStore);
```
Java will use the cacerts bundle provided by the jdk installation in that scenario.

bjorncs commented on security-utils/src/main/java/com/yahoo/security/tls/json/TransportSecurityOptionsJsonSerializer.java:
```diff
@@ -0,0 +1,143 @@
+
+    private static PeerPolicy toPeerPolicy(AuthorizedPeer authorizedPeer) {
+        if (authorizedPeer.name == null) {
+            throw missingFieldException("name");
```
👍

bjorncs committed:
Rename 'allowed-peers' to 'authorized-peers'
